kernel/src/grant.rs

```diff
@@ -83,22 +83,32 @@ impl<T: ?Sized> DerefMut for Owned<T> {
 impl Allocator {
     pub fn alloc<T>(&mut self, data: T) -> Result<Owned<T>, Error> {
         unsafe {
-            self.appid
-                .kernel
-                .process_map_or(Err(Error::NoSuchApp), self.appid, |process| {
-                    process.alloc(size_of::<T>(), align_of::<T>()).map_or(
-                        Err(Error::OutOfMemory),
-                        |arr| {
-                            let ptr = arr.as_mut_ptr() as *mut T;
-                            // We use `ptr::write` to avoid `Drop`ping the uninitialized memory in
-                            // case `T` implements the `Drop` trait.
-                            write(ptr, data);
-                            Ok(Owned::new(ptr, self.appid))
-                        },
-                    )
-                })
+            let ptr = self.alloc_unowned(data)?;
+            Ok(Owned::new(ptr, self.appid))
         }
     }
+
+    // Like `alloc`, but the caller is responsible for free-ing the allocated
+    // memory, as it is not wrapped in a type that implements `Drop`
+    unsafe fn alloc_unowned<T>(&mut self, data: T) -> Result<NonNull<T>, Error> {
+        self.appid
+            .kernel
+            .process_map_or(Err(Error::NoSuchApp), self.appid, |process| {
+                process.alloc(size_of::<T>(), align_of::<T>()).map_or(
+                    Err(Error::OutOfMemory),
+                    |buf| {
+                        // Convert untyped `*mut u8` allocation to allocated type
+                        let ptr = NonNull::new_unchecked(buf.as_ptr() as *mut T);
```

Comment:
This `new_unchecked` is unnecessary. At this point `buf` is already a `NonNull<u8>`.

Reply:
Btw, wouldn't it be simpler to have the `process::alloc` function be `fn alloc<T: Sized>(&self) -> Option<NonNull<T>>`, which would internally compute `sizeof<T>` and `alignof<T>` and convert back to a `NonNull<T>`, instead of the current untyped function working on raw `usize` and `NonNull<u8>`?

Reply:
The goal of that line was to convert a `NonNull<u8>` to a `NonNull<T>`,  which I believe you can only do by manipulating the underlying pointer -- however, pushing the type into `alloc` makes more sense to me, so I'll look into that.

Reply:
Okay, I went down this road a little bit, but you run into the problem that nothing in the `ProcessType` trait is currently generic:

```
error[E0038]: the trait `process::ProcessType` cannot be made into an object
   --> tock/kernel/src/sched.rs:327:18
    |
327 |         process: &dyn process::ProcessType,
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `process::ProcessType` cannot be made into an object
    |
   ::: tock/kernel/src/process.rs:183:11
    |
183 | pub trait ProcessType {
    |           ----------- this trait cannot be made into an object...
...
350 |     fn alloc<T: Sized>(&self) -> Option<NonNull<T>>;
    |        ----- ...because method `alloc` has generic type parameters
    |
    = help: consider moving `alloc` to another trait
```

Fixing this touches a lot of code, and I think expands the scope of this PR too much.

Reply:
Original line replaced by `cast()`